### PR TITLE
Perform only single protocol dispatching on Enum.fetch/2

### PR DESCRIPTION
**Pull request by michalmuskala**
Extracted from https://github.com/elixir-lang/elixir/pull/5818

### CODE REVIEW

File: lib/elixir/lib/enum.ex [hunk 1]
```diff
@@ -723,28 +723,30 @@ defmodule Enum do
   end
 
   def fetch(enumerable, index) when is_integer(index) and index < 0 do
-    case Enumerable.count(enumerable) do
+    module = Enumerable.impl_for!(enumerable)
+    case module.count(enumerable) do
       {:error, _module} ->
         enumerable |> reverse |> fetch_list((-index) - 1)
```
Comment (michalmuskala) [inline lib/elixir/lib/enum.ex]: Hmm... It seems we're still going to do a second dispatch in this branch on reverse.

Comment (josevalim) [inline lib/elixir/lib/enum.ex]: Yes, we need to implement reverse by hand:

    enumerable |> module.reverse({:cont, []}, fn x, acc -> {:cont, [x | acc]} end) |> elem(1) |> fetch_list((-index) - 1)

File: lib/elixir/lib/enum.ex [hunk 2]
```diff
@@ -723,28 +723,29 @@ defmodule Enum do
...
   def fetch(enumerable, index) when is_integer(index) and index < 0 do
-    case Enumerable.count(enumerable) do
-      {:error, _module} ->
-        enumerable |> reverse |> fetch_list((-index) - 1)
-
+    module = Enumerable.impl_for!(enumerable)
+    case module.count(enumerable) do
+      {:error, module} ->
+        reversed =
+          module.reduce(enumerable, {:cont, []},
```
Comment (lexmag) [inline lib/elixir/lib/enum.ex]: The following style would be more common:
```elixir
module.reduce(enumerable, {:cont, []}, fn item, acc ->
  {:cont, [item | acc]}
end) |> elem(1)
```
And let's not use `x`.

Comment (michalmuskala) [inline lib/elixir/lib/enum.ex]: :+1: I copied from the `revert/1` implementation

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Pull request merged**